alfalfa_worker/worker_fmu/step_fmu.py

```diff
@@ -164,13 +169,41 @@ def run(self):
                         self.set_idle_state()
                         break
         else:
-            while self.simtime < self.endTime:
-                if self.db_stop_set():
-                    break
+            # first step takes extra time and needs to happen outside timescale loop
+            if self.simtime == self.startTime:
+                print("taking first step at ", datetime.now())
                 self.step()
-                # TODO: Make this respect time scale provided by user
-                time.sleep(5)
+                print("finished first step at ", datetime.now())
+                next
+            current_time = datetime.now()
+            next_step_time = current_time + self.realworld_timedelta
+            print("in run with current time & next_step_time: ", current_time, next_step_time)
+            self.advance = False
+            while True:
+                current_time = datetime.now()
+
+                if current_time >= next_step_time:
+                    self.advance = True
+                    # sim is not keeping up with target timescale.
+                    # TODO rethink arbitrary 60s behind
+                    if (current_time > next_step_time + timedelta(seconds=60)):
+                        self.stop = True
+                        print("stopping... simulation got more than 60s behind target timescale")
+
+                if self.stop:
+                    break
+
+                if self.advance:
+                    self.step()
+                    next_step_time = next_step_time + self.realworld_timedelta
+                    self.advance = False
+
+                # update stop flag from db
+                self.db_stop_set()
 
+                # update stop flag from endTime
+                if self.simtime >= self.endTime:
```

Comment:
Should this check and the `db_stop_set()` happen before the `if self.stop: break` block? That way you're guaranteed that you won't have a step run after the simulation is supposed to stop.

Reply:
Yes again.

Reply:
was any code supposed to be pushed to respond to this?

Reply:
I just pushed another update.  Since the step() is what updates self.simtime, I don't think this makes a difference functionally but I do think it's cleaner and more readable.